### PR TITLE
Added support for trace calculations, trace feeders, and check feeders

### PR DESCRIPTION
**Pull request by rclapp**
New methods for calculation tracing.

### CODE REVIEW

**Comment by rclapp** [comment]
https://github.com/cubewise-code/tm1py/issues/779

**Comment by MariusWirtz** [comment]
Thanks for the contribution! I will review and merge in the next few days

File: TM1py/Services/CellService.py [hunk 1]
```diff
@@ -238,6 +237,168 @@ def get_value(self, cube_name: str, element_string: str, dimensions: List[str] =
...
+                         "?$select=Fed"
+                         "&$expand=Tuple($select=Name,UniqueName,Type),Cube($select=Name)", cube_name)
+
+        url = add_url_parameters(url, **{"!sandbox": sandbox_name})
+        if isinstance(elements, str):
+            body_as_dict = self._compose_odata_tuple_from_string(cube_name, elements, dimensions)
+        else:
+            body_as_dict = self._compose_odata_tuple_from_iterable(cube_name, elements, dimensions)
+        data = json.dumps(body_as_dict, ensure_ascii=False)
+        return self._rest.POST(url=url, data=data, **kwargs).content
```
Comment (MariusWirtz) [inline TM1py/Services/CellService.py:399]: In most existing functions we return a dictionary (`response.json()`). 
Why choose a str here instead?

Comment (rclapp) [reply TM1py/Services/CellService.py:399]: TBH, I had a reason when I started this, but now now I don't recall what it was. I can change it to a dict.

File: TM1py/Services/CellService.py [hunk 2]
```diff
@@ -593,6 +757,8 @@ async def _write_async(df: 'pd.DataFrame'):
             error_log_files=list(itertools.chain(*[exception.error_log_files for exception in exceptions])),
             attempts=sum([exception.attempts for exception in exceptions]))
 
+
+    @manage_changeset
```
Comment (MariusWirtz) [inline TM1py/Services/CellService.py:761]: Nice one. good catch

File: TM1py/Services/CellService.py [hunk 3]
```diff
@@ -188,9 +188,8 @@ def get_value(self, cube_name: str, element_string: str, dimensions: List[str] =
         """
         mdx_template = "SELECT {} ON ROWS, {} ON COLUMNS FROM [{}]"
         mdx_rows_list = []
-        from TM1py.Services.CubeService import CubeService
         if not dimensions:
-            dimensions = CubeService(self._rest).get(cube_name).dimensions
+            dimensions = self.get_dimension_names_for_writing(cube_name=cube_name)
```
Comment (MariusWirtz) [inline TM1py/Services/CellService.py:192]: Good Catch. I think the previous code would have failed in a model with sandboxes as a first dimension.

**Comment by MariusWirtz** [comment]
@rclapp 
I am not sure if the `trace_cell_feeders` function works.
for me it returns an empty response from a cell that is fed.

Am I missing something? Can you please share a working sample?

```python
from TM1py import TM1Service

with TM1Service(address='localhost', port=12354, ssl=True, user="admin", password="apple") as tm1:
    cell_trace = tm1.cells.trace_cell_feeders(cube_name="c1", elements="e1,e1")
    print(cell_trace)

    something = tm1.cells.check_cell_feeders(cube_name="c1", elements="e1,e1")
    print(something)
```

```
{'@odata.context': '../$metadata#ibm.tm1.api.v1.FeederTrace', 'FedCells': [], 'Statements': []}
{'@odata.context': '../$metadata#Collection(ibm.tm1.api.v1.FedCellDescriptor)', 'value': []}
```

image

**Comment by rclapp** [comment]
I am opening a case with IBM about it.

From: Marius Wirtz ***@***.***>
Reply-To: cubewise-code/tm1py ***@***.***>
Date: Saturday, August 20, 2022 at 3:42 PM
To: cubewise-code/tm1py ***@***.***>
Cc: "Clapp, Ryan" ***@***.***>, Mention ***@***.***>
Subject: Re: [cubewise-code/tm1py] Added support for trace calculations, trace feeders, and check feeders (PR #780)


@rclapp<https://github.com/rclapp>
I am not sure if the trace_cell_feeders function works.
for me it returns an empty response from a cell that is fed.

Am I missing something? Can you please share a working sample?

from TM1py import TM1Service



with TM1Service(address='localhost', port=12354, ssl=True, user="admin", password="apple") as tm1:

    cell_trace = tm1.cells.trace_cell_feeders(cube_name="c1", elements="e1,e1")

    print(cell_trace)



    something = tm1.cells.check_cell_feeders(cube_name="c1", elements="e1,e1")

    print(something)

***@***.***': '../$metadata#ibm.tm1.api.v1.FeederTrace', 'FedCells': [], 'Statements': []}

***@***.***': '../$metadata#Collection(ibm.tm1.api.v1.FedCellDescriptor)', 'value': []}

[Image removed by sender. image]<https://user-images.githubusercontent.com/15182131/185763595-05b4fe14-5438-4249-8bce-eb49cbcef5a9.png>

**Pull request merged**